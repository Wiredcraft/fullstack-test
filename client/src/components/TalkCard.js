import React, { useContext } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import UserContext from '../components/UserContext';
import { apiHost } from '../utils/config';
import ordinalize from '../utils/ordinalize';
import { fireModal, NOT_LOGGED_IN, NETWORK_ERROR } from '../utils/fireModal';

const TalkCard = ({ idx, setActive, votes, updateVotes, title, abstract, id, active }, ref) => {
  const { user } = useContext(UserContext);

  const vote = async val => {
    if (!user) {
      fireModal(NOT_LOGGED_IN, {actionName: 'vote'});
    } else {
      try {
        const res = await fetch(`${apiHost}/talks/${id}/${val}`, {
          method: 'POST',
          headers: {
            'Authorization': 'Bearer ' + user
          }
        });
        const json = await res.json();

        updateVotes(json.newVal);
      } catch(e) {
        fireModal(NETWORK_ERROR);
      }
    }
  };

  const medal = [ 'gold', 'silver', 'bronze' ][idx];

  const toggle = () => active ? setActive(null) : setActive(id);

  const toggleOnEnter = e => {
    if (e.target === e.currentTarget // only fire for outer div, not inner controls
      && e.key === 'Enter') {
      toggle();
    }
  };

  return (
    <div
      className={active ? 'active talk' : 'talk'}
      tabIndex={0}
      onClick={() => setActive(id)}
      onKeyPress={toggleOnEnter}
      ref={active ? ref : null}
    >
      <div>
        <div className='header-row'>
          {medal ? (
            <FontAwesomeIcon
              className={`medal ${medal}`}
              icon='medal'
              aria-label={`${ordinalize(idx + 1)} place`}
              title={`${ordinalize(idx + 1)} place`}
            />
          ) : <span
              className='place'
              aria-label={`${ordinalize(idx + 1)} place`}
              title={`${ordinalize(idx + 1)} place`}
            >
              {idx + 1}
            </span>}

          <h3>{title}</h3>
          <div className='votes'>
            Votes: {votes}
            {
              active && (
                <React.Fragment>
                  <button className='unstyled' onClick={() => vote('upvote')}>
                    <FontAwesomeIcon
                      className='vote-btn upvote'
                      icon='thumbs-up'
                      aria-label='upvote'
                      title='upvote'
                    />
                  </button>
                  <button className='unstyled' onClick={() => vote('downvote')}>
                    <FontAwesomeIcon
                      className='vote-btn downvote'
                      icon='thumbs-down'
                      aria-label='downvote'
                      title='downvote'
                    />
                  </button>
                </React.Fragment>
              )
            }
          </div>
        </div>
      </div>
      {active && (
        <React.Fragment>{
          abstract
            .split(/\n\s*\n/)
            .map((para, idx) => <p key={idx}>{para}</p>)
        }</React.Fragment>
      )}
    </div>
  );
};

export default React.forwardRef(TalkCard);
