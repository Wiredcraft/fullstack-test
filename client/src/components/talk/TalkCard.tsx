import * as React from "react";
import { useTalks } from "../../contexts/TalksContext";
import { getTalks, voteTalk } from "../../services/talks";
import { TalkCardInfo } from '../../types/talk';
import "./talkCard.css";

type InputProps = {
  isAuthenticated: boolean,
  cardInfo: TalkCardInfo
};

function TalkCard(props: InputProps) {
  const { cardInfo: {
    _id: id,
    author,
    description,
    title,
    voteCount,
    votedByUser,
    createdAt
  },      isAuthenticated } = props;

  const { reloadTalks } = useTalks();

  const operation = votedByUser ? "REMOVE" : "ADD";
  const buttonClass = votedByUser ? "talk-card-container__button--voted" : "talk-card-container__button--not-voted";
  const createdDate = new Date(createdAt!.toString());
  const formattedDate = `${createdDate.getHours()} : ${createdDate.getMinutes()}, ${createdDate.toDateString()}`;

  const onVote = async () => {
    voteTalk({ talkId: id!, operation });
    const talksFromServer = await getTalks();
    talksFromServer.length ? reloadTalks(talksFromServer) : null;
  };

  return (
    <div className="talk-card-container">
      {
        isAuthenticated ? <button
          onClick={onVote}
          className={buttonClass}>{voteCount}</button>
          : null
      }
      <div className="talk-card-container--row">
        <h1>{title}</h1>
        <p>{description}</p>
        <h3>by: {author}</h3>
        <p className="talk-created-date">{formattedDate}</p>
      </div>
    </div>
  );
}

export default TalkCard;
