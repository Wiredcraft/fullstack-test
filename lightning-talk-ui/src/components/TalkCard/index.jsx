import React, { useCallback, useMemo } from 'react';
import PropTypes from 'prop-types';
import { BASE_URL, like } from '@/request';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

import Player from '@/components/Player';

const TalkCard = (props) => {
  const {
    id,
    title,
    video,
    history,
    description,
    fetchTalks,
    user,
    users,
  } = props;

  const upVote = useCallback(
    async () => {
      if (!user) {
        history.push('/login');
        return;
      }
      try {
        await like(id, { users });
        fetchTalks();
      } catch (e) {
        if (e.statusCode && e.statusCode === 403) {
          history.push('/login');
        }
      }
    },
    [id, user, users],
  );

  const videos = video.map((v) => {
    const url = new URL(v.url, BASE_URL);
    return <Player key={v.id} src={url.href} />;
  });

  const liked = useMemo(
    () => !!users.find((u) => u.id === user),
    [users, user],
  );

  return (
    <div className="talk">
      <div className="talk__title">
        <h3>{title}</h3>
      </div>
      <div className="talk__player">{videos}</div>
      <div className="talk__desc">
        <p>{description}</p>
      </div>
      <button
        type="button"
        className={`talk__polling btn--text ${liked ? 'liked' : ''}`}
        onClick={upVote}
      >
        <span className="talk__polling-num">{users.length}</span>
      </button>
    </div>
  );
};

TalkCard.propTypes = {
  id: PropTypes.number.isRequired,
  user: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]).isRequired,
  title: PropTypes.string.isRequired,
  users: PropTypes.array.isRequired,
  description: PropTypes.string.isRequired,
  video: PropTypes.arrayOf(PropTypes.object).isRequired,
  history: PropTypes.object.isRequired,
  fetchTalks: PropTypes.func.isRequired,
};

TalkCard.defaultProps = {
};

const mapState = (state) => ({
  user: state.user.user,
});

const mapDispatch = ({ talks: { fetchTalks } }) => ({
  fetchTalks,
});

export default connect(
  mapState,
  mapDispatch,
)(withRouter(TalkCard));
