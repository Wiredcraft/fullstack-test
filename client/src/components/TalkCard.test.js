import React from 'react';
import { render } from 'react-testing-library';
import TalkCard from './TalkCard';

it('renders TalkCard correctly', () => {
  const { asFragment } = render(<TalkCard />);

  expect(asFragment()).toMatchSnapshot();
});
