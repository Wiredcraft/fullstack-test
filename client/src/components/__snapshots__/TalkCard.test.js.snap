// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TalkCard correctly 1`] = `
<DocumentFragment>
  <div
    class="talk"
    tabindex="0"
  >
    <div>
      <div
        class="header-row"
      >
        <h3 />
        <div
          class="votes"
        >
          Votes: 
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
